File: Week_02/java/src/_99_RecoverBinarySearchTree.java
import java.util.Deque;
import java.util.LinkedList;

/**
 * 第一遍：2020/08/08周六 ✅
 * 第二遍：2020/08/09周日 ✅
 * 第三遍：2020/08/11周二 ✅
 * 第三遍：2020/07/08周四
 * 抄起来挺简单，但我自己想不出来啊。。。
 */
class _99_RecoverBinarySearchTree {
    class SolutionMorris {
        public void recoverTree(TreeNode root) {
            TreeNode x = null, y = null, pre = null, predecessor = null, node = root;
            while (node != null) {
                if (node.left != null) {
                    predecessor = node.left;
                    while (predecessor.right != null && predecessor.right != node) {
                        predecessor = predecessor.right;
                    }

                    if (predecessor.right == null) {
                        predecessor.right = node;
                        node = node.left;
                    } else {
                        if (pre != null && pre.val > node.val) {
                            y = node;
                            if (x == null) {
                                x = pre;
                            }
                        }
                        predecessor.right = null;
                        pre = node;
                        node = node.right;
                    }
                } else {
                    if (pre != null && pre.val > node.val) {
                        y = node;
                        if (x == null) {
                            x = pre;
                        }
                    }
                    pre = node;
                    node = node.right;
                }
            }

            int temp = x.val;
            x.val = y.val;
            y.val = temp;
        }
    }

    public void recoverTree(TreeNode root) {
        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode pre = null, x = null, y = null, node = root;
        while (node != null || !stack.isEmpty()) {
            while (node != null) {
                stack.push(node);
                node = node.left;
            }
            node = stack.pop();
            if (pre != null && pre.val > node.val) {
                y = node;
                if (x == null) {
                    x = pre;
                } else {
                    break;
                }
            }
            pre = node;
            node = node.right;
        }
        int temp = x.val;
        x.val = y.val;
        y.val = temp;
    }

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}

File: Week_02/dart/src/99_RecoverBinarySearchTree.dart
import 'dart:collection';

/**
 * Definition for a binary tree node.

 */
class Solution {
  void recoverTree2(TreeNode? root) {
    TreeNode? x, y, pre;

    final stack = Queue<TreeNode>();
    while (root != null || stack.isNotEmpty) {
      while (root != null) {
        stack.addLast(root);
        root = root.left;
      }
      final node = stack.removeLast();
      if (pre != null && pre.val > node.val) {
        y = node;
        if (x == null) {
          x = pre;
        }
      }
      pre = node;
      root = node.right;
    }

    if (x != null && y != null) {
      final t = x.val;
      x.val = y.val;
      y.val = t;
    }
  }

  void recoverTree(TreeNode? root) {
    TreeNode? x, y, pre;
    while (root != null) {
      if (root.left != null) {
        var p = root.left;
        while (p?.right != null && p?.right != root) {
          p = p?.right;
        }
        if (p?.right == null) {
          p?.right = root;
          root = root.left;
        } else {
          if (pre != null && pre.val > root.val) {
            y = root;
            if (x == null) {
              x = pre;
            }
          }
          p?.right = null;
          pre = root;
          root = root.right;
        }
      } else {
        if (pre != null && pre.val > root.val) {
          y = root;
          if (x == null) {
            x = pre;
          }
        }
        pre = root;
        root = root.right;
      }
    }
    if (x != null && y != null) {
      final t = x.val;
      x.val = y.val;
      y.val = t;
    }
  }
}

class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;
  TreeNode([this.val = 0, this.left, this.right]);
}
